mentat/commands.py

```diff
@@ -0,0 +1,104 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from typing import Iterable
+
+from termcolor import cprint
+
+from .errors import MentatError
+from .git_handler import commit
+
+
+class Command(ABC):
+    _registered_commands = {}
+
+    def __init_subclass__(cls, command_name: str | None) -> None:
+        if command_name is not None:
+            Command._registered_commands[command_name] = cls
+
+    @classmethod
+    def create_command(cls, command_name: str) -> Command:
+        if command_name not in cls._registered_commands:
+            return InvalidCommand(command_name)
+        else:
+            return cls._registered_commands[command_name]()
+
+    @classmethod
+    def get_command_completions(cls) -> Iterable[str]:
+        return list(map(lambda name: "/" + name, cls._registered_commands))
+
+    # *args[0] is always the command_name, c style
+    @abstractmethod
+    def apply(self, *args: str) -> None:
+        pass
+
+    # TODO: make more robust way to specify arguments for commands
+    @classmethod
+    @abstractmethod
+    def argument_names(cls) -> list[str]:
+        pass
+
+    @classmethod
+    @abstractmethod
+    def help_message(cls) -> str:
+        pass
+
+
+class InvalidCommand(Command, command_name=None):
+    def __init__(self, invalid_name):
+        self.invalid_name = invalid_name
+
+    def apply(self, *args: str) -> None:
+        cprint(
+            f"{self.invalid_name} is not a valid command. Use /help to see a list of"
+            " all valid commands",
+            color="light_yellow",
+        )
+
+    @classmethod
+    def argument_names(cls) -> list[str]:
+        raise MentatError("Argument names called on invalid command")
+
+    @classmethod
+    def help_message(cls) -> str:
+        raise MentatError("Help message called on invalid command")
+
+
+help_message_width = 60
+
+
+class HelpCommand(Command, command_name="help"):
+    def apply(self, *args: str) -> None:
+        for command_name, command_class in Command._registered_commands.items():
+            argument_names = command_class.argument_names()
+            help_message = command_class.help_message()
+            message = " ".join(
+                [f"/{command_name}"] + [f"<{arg}>" for arg in argument_names]
+            )
+            print(message.ljust(help_message_width), help_message, sep="")
+
+    @classmethod
+    def argument_names(cls) -> list[str]:
+        return []
+
+    @classmethod
+    def help_message(cls) -> str:
+        return "Displays this message"
+
+
+class CommitCommand(Command, command_name="commit"):
+    default_message = "Automatic commit"
+
+    def apply(self, *args: str) -> None:
+        if len(args) > 1:
+            commit(args[1])
```

Comment:
most commit messages will be multiple words separated by a space. On the command line you'd do `git commit -m "multiple words separated by spaces"`, so we should probably also capture everything in quotes

Reply:
Hmm, good point; I'm wondering if there's a library that can do argument parsing for us? I'll look into this

Reply:
It looks like python has a builtin library shlex.split that will automatically handle this all for us!